src/oracles/ChainlinkBoundedCompositeOracle.sol

```diff
@@ -0,0 +1,275 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+import {SafeCast} from "@openzeppelin-contracts/contracts/utils/math/SafeCast.sol";
+import {Ownable} from "@openzeppelin-contracts/contracts/access/Ownable.sol";
+import {Address} from "@openzeppelin-contracts/contracts/utils/Address.sol";
+
+import {AggregatorV3Interface} from "./AggregatorV3Interface.sol";
+
+/// @title ChainlinkBoundedCompositeOracle
+/// @notice Oracle that checks if a primary price feed is within bounds, then either:
+/// - If within bounds: Combines it with a second price feed
+/// - If out of bounds: Falls back to a backup price feed
+/// @dev Includes OEV (Oracle Extractable Value) functionality for early updates
+contract ChainlinkBoundedCompositeOracle is AggregatorV3Interface, Ownable {
+    using Address for address;
+    using SafeCast for *;
+
+    /// @notice Emitted when bounds are updated
+    event BoundsUpdated(
+        int256 oldLower,
+        int256 newLower,
+        int256 oldUpper,
+        int256 newUpper
+    );
+
+    /// @notice Emitted when the fee multiplier is updated
+    event FeeMultiplierUpdated(
+        uint16 oldFeeMultiplier,
+        uint16 newFeeMultiplier
+    );
+
+    /// @notice Emitted when the early update window is updated
+    event EarlyUpdateWindowUpdated(uint256 oldWindow, uint256 newWindow);
+
+    /// @notice Emitted when primary oracle address is updated
+    event PrimaryOracleUpdated(
+        address oldPrimaryOracle,
+        address newPrimaryOracle
+    );
+
+    /// @notice Emitted when BTC oracle address is updated
+    event BTCOracleUpdated(address oldBTCOracle, address newBTCOracle);
+
+    /// @notice Emitted when fallback oracle address is updated
+    event FallbackOracleUpdated(
+        address oldFallbackOracle,
+        address newFallbackOracle
+    );
+
+    /// @notice Primary oracle to check bounds against
+    AggregatorV3Interface public primaryLBTCOracle;
+
+    /// @notice Fallback market rate oracle to use if primary is out of bounds
+    AggregatorV3Interface public fallbackLBTCOracle;
+
+    /// @notice Lower bound for primary oracle price
+    int256 public lowerBound;
+
+    /// @notice Upper bound for primary oracle price
+    int256 public upperBound;
+
+    /// @notice Scaling factor for price calculations
+    uint8 public constant decimals = 18;
```

Comment:
my mistake, it should be 8, to match redstone's PoR.
it would also work with 18, but then we might get confused with lower/upper bound configuration, as they will have to be with 18 decimals and not 8.